### PR TITLE
[FEATURE] `query.pair_column` Metric

### PR DESCRIPTION
**Pull request by austiezr**
Please annotate your PR title to describe what the PR does, then give a brief bulleted description of your PR below. PR titles should begin with [BUGFIX], [FEATURE],  [DOCS], or [MAINTENANCE]. If a new feature introduces breaking changes for the Great Expectations API or configuration files, please also add [BREAKING]. You can read about the tags in our [contributor checklist](https://docs.greatexpectations.io/docs/contributing/contributing_checklist).

Changes proposed in this pull request:
- Adds a `query.pair_column` metric to support Custom Query PairColumn Expectations
- Adds a contrib expectation as an example using `query.pair_column`

After submitting your PR, CI checks will run and @cla-bot will check for your CLA signature.

For a PR with nontrivial changes, we review with both design-centric and code-centric lenses.

In a design review, we aim to ensure that the PR is consistent with our relationship to the open source community, with our software architecture and abstractions, and with our users' needs and expectations. That review often starts well before a PR, for example in github issues or slack, so please link to relevant conversations in notes below to help reviewers understand and approve your PR more quickly (e.g. `closes #123`).

Previous Design Review notes:
- Extends #5223 

### Definition of Done
Please delete options that are not relevant.

- [x] My code follows the Great Expectations [style guide](https://docs.greatexpectations.io/docs/contributing/style_guides/code_style)
- [x] I have performed a [self-review](https://docs.greatexpectations.io/docs/contributing/contributing_checklist) of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have run any local integration tests and made sure that nothing is broken.

Thank you for submitting!

### CODE REVIEW

**Comment by austiezr** [comment]
@moqidimoc 👀 🚀

**Review by AFineDayFor — APPROVED** [review]
lgtm

**Review by rdodev — APPROVED** [review]
LGTM pending mario 👀

File: contrib/experimental/great_expectations_experimental/expectations/expect_queried_pair_column_values_to_have_diff.py [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+    InvalidExpectationConfigurationError,
+)
+from great_expectations.execution_engine import ExecutionEngine
+from great_expectations.expectations.expectation import (
+    ExpectationValidationResult,
+    QueryExpectation,
+)
+
+
+class ExpectQueriedPairColumnValuesToHaveDiff(QueryExpectation):
```
Comment (alexsherstinsky) [inline contrib/experimental/great_expectations_experimental/expectations/expect_queried_pair_column_values_to_have_diff.py]: @austiezr May I please propose the naming change from PairColumn to ColumnPair throughout -- so as to be consistent with the naming of similar entities in the core sections of the code?  Thanks!

File: contrib/experimental/great_expectations_experimental/expectations/expect_queried_pair_column_values_to_have_diff.py [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+    """Expect the frequency of occurrences of a specified value in a queried column to be at least <threshold> percent of values in that column."""
```
Comment (alexsherstinsky) [inline contrib/experimental/great_expectations_experimental/expectations/expect_queried_pair_column_values_to_have_diff.py]: @austiezr Which success kwarg is the threshold?  This is not entirely clear to me -- thanks.

File: contrib/experimental/great_expectations_experimental/expectations/expect_queried_pair_column_values_to_have_diff.py [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+            raise InvalidExpectationConfigurationError(str(e))
+
+    def _validate(
+        self,
+        configuration: ExpectationConfiguration,
+        metrics: dict,
+        runtime_configuration: dict = None,
+        execution_engine: ExecutionEngine = None,
+    ) -> Union[ExpectationValidationResult, dict]:
+
```
Comment (alexsherstinsky) [inline contrib/experimental/great_expectations_experimental/expectations/expect_queried_pair_column_values_to_have_diff.py]: @austiezr Let us delete this blank line (style).  Thanks.

File: contrib/experimental/great_expectations_experimental/expectations/expect_queried_pair_column_values_to_have_diff.py [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+        diff = configuration["kwargs"].get("diff")
+        mostly = configuration["kwargs"].get("mostly")
+        strict = configuration["kwargs"].get("strict")
+        query_result = metrics.get("query.pair_column")
```
Comment (alexsherstinsky) [inline contrib/experimental/great_expectations_experimental/expectations/expect_queried_pair_column_values_to_have_diff.py]: @austiezr Again, hope it is possible to change the name from "query.pair_column" to "query.column_pair".  Thanks.

File: contrib/experimental/great_expectations_experimental/expectations/expect_queried_pair_column_values_to_have_diff.py [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+
+        if mostly == 1:
```
Comment (alexsherstinsky) [inline contrib/experimental/great_expectations_experimental/expectations/expect_queried_pair_column_values_to_have_diff.py]: @austiezr I feel that this `if` statement can be shortened by combining all cases for each value of `strict` into one expression.  Could you please give it a try and see if the tests still pass?  Thanks!

File: contrib/experimental/great_expectations_experimental/expectations/expect_queried_pair_column_values_to_have_diff.py [hunk 6]
```diff
@@ -0,0 +1,142 @@
...
+            success = (
+                sum([(abs(x[0]) == diff) for x in query_result]) / len(query_result)
+            ) > mostly
+        else:
+            success = (
+                sum([(abs(x[0]) == diff) for x in query_result]) / len(query_result)
+            ) >= mostly
+
+        return {
+            "success": success,
```
Comment (alexsherstinsky) [inline contrib/experimental/great_expectations_experimental/expectations/expect_queried_pair_column_values_to_have_diff.py]: @austiezr Question: is `x` a list of length 1?  Or are there other elements in each `x`?  Thanks!

File: contrib/experimental/great_expectations_experimental/expectations/expect_queried_pair_column_values_to_have_diff.py [hunk 7]
```diff
@@ -0,0 +1,142 @@
...
+                    "out": {"success": True},
+                    "only_for": ["sqlite", "spark"],
+                },
+                {
+                    "title": "basic_negative_test",
+                    "exact_match_out": False,
+                    "include_in_gallery": True,
+                    "in": {
+                        "column_A": "col1",
+                        "column_B": "col2",
```
Comment (alexsherstinsky) [inline contrib/experimental/great_expectations_experimental/expectations/expect_queried_pair_column_values_to_have_diff.py]: @austiezr Should we add some tests that are floating point numbers?  Are there checks against non-applicable types in the column?  Thanks.

File: great_expectations/expectations/metrics/query_metrics/query_pair_column.py [hunk 8]
```diff
@@ -0,0 +1,107 @@
+from typing import Any, Dict, List, Optional, Union
+
```
Comment (alexsherstinsky) [inline great_expectations/expectations/metrics/query_metrics/query_pair_column.py]: @austiezr Again, are we able to rename this module to `query_column_pair.py`?  Thanks.

File: great_expectations/expectations/metrics/query_metrics/query_pair_column.py [hunk 9]
```diff
@@ -0,0 +1,107 @@
...
+    sqlalchemy_engine_Row,
+)
+from great_expectations.expectations.metrics.metric_provider import metric_value
+from great_expectations.expectations.metrics.query_metric_provider import (
+    QueryMetricProvider,
+)
+from great_expectations.util import get_sqlalchemy_subquery_type
+
+
+class QueryPairColumn(QueryMetricProvider):
```
Comment (alexsherstinsky) [inline great_expectations/expectations/metrics/query_metrics/query_pair_column.py]: @austiezr Asking for a name change as per above.  Thanks!

File: great_expectations/expectations/metrics/query_metrics/query_pair_column.py [hunk 10]
```diff
@@ -0,0 +1,107 @@
...
+    metric_name = "query.pair_column"
```
Comment (alexsherstinsky) [inline great_expectations/expectations/metrics/query_metrics/query_pair_column.py]: @austiezr Same request -- metric name change.  Thanks!

File: great_expectations/expectations/metrics/query_metrics/query_pair_column.py [hunk 11]
```diff
@@ -0,0 +1,107 @@
...
+    value_keys = (
+        "column_A",
+        "column_B",
+        "query",
+    )
+
+    @metric_value(engine=SqlAlchemyExecutionEngine)
+    def _sqlalchemy(
```
Comment (alexsherstinsky) [inline great_expectations/expectations/metrics/query_metrics/query_pair_column.py]: ❤️

File: great_expectations/expectations/metrics/query_metrics/query_pair_column.py [hunk 12]
```diff
@@ -0,0 +1,107 @@
...
+            query = query.format(
+                column_A=column_A, column_B=column_B, active_batch=f"({selectable})"
+            )
+
+        engine: sqlalchemy_engine_Engine = execution_engine.engine
+        result: List[sqlalchemy_engine_Row] = engine.execute(sa.text(query)).fetchall()
+
+        return result
+
+    @metric_value(engine=SparkDFExecutionEngine)
```
Comment (alexsherstinsky) [inline great_expectations/expectations/metrics/query_metrics/query_pair_column.py]: ❤️

**Review by alexsherstinsky — CHANGES_REQUESTED** [review]
@austiezr I love the core implementation -- just asking for a few stylistic refinements.  Happy to discuss.  Thank you!

File: contrib/experimental/great_expectations_experimental/expectations/expect_queried_pair_column_values_to_have_diff.py (hunk 6 above)
Comment (austiezr) [reply contrib/experimental/great_expectations_experimental/expectations/expect_queried_pair_column_values_to_have_diff.py]: Similar; `x` in this context a `Row` object of some kind, containing a single value being indexed.

File: contrib/experimental/great_expectations_experimental/expectations/expect_queried_column_pair_values_to_have_diff.py [hunk 13]
```diff
@@ -0,0 +1,176 @@
...
+        execution_engine: ExecutionEngine = None,
+    ) -> Union[ExpectationValidationResult, dict]:
+        diff: Union[float, int] = configuration["kwargs"].get("diff")
+        mostly: str = configuration["kwargs"].get("mostly")
+        strict: bool = configuration["kwargs"].get("strict")
+        query_result: Union[sqlalchemy_engine_Row, pyspark_sql_Row] = metrics.get(
+            "query.column_pair"
+        )
+
+        if mostly != 1 and strict is True:
```
Comment (alexsherstinsky) [inline contrib/experimental/great_expectations_experimental/expectations/expect_queried_column_pair_values_to_have_diff.py]: @austiezr Sorry, I was hoping that the value of `mostly` being `1` will not need to be handled in any special way.  I believe that it should be possible to let everything be handled by `>` and `>=`, regardless of the value of `mostly`.  Do you agree?  Happy to discuss.  Thanks!

**Review by alexsherstinsky — CHANGES_REQUESTED** [review]
@austiezr One more small change request -- thanks!

File: contrib/experimental/great_expectations_experimental/expectations/expect_queried_column_pair_values_to_have_diff.py (hunk 13 above)
Comment (austiezr) [reply contrib/experimental/great_expectations_experimental/expectations/expect_queried_column_pair_values_to_have_diff.py]: Hey @alexsherstinsky ! Agree; realized I was still working with a `strict` parameter held over from another template. Simplified now, tests pass, should be good to go!

**Review by alexsherstinsky — APPROVED** [review]
LGTM -- @austiezr -- Yay!!!  Thank you!

**Pull request merged**